core/src/main/scala/kafka/server/AdminManager.scala

```diff
@@ -255,14 +258,28 @@ class AdminManager(val config: KafkaConfig,
         resource.`type` match {
           case ResourceType.TOPIC =>
             val topic = resource.name
+
             val properties = new Properties
             config.entries.asScala.foreach { configEntry =>
               properties.setProperty(configEntry.name(), configEntry.value())
             }
-            if (validateOnly)
-              AdminUtils.validateTopicConfig(zkUtils, topic, properties)
-            else
-              AdminUtils.changeTopicConfig(zkUtils, topic, properties)
+
+            alterConfigPolicy match {
+              case Some(policy) =>
+                AdminUtils.validateTopicConfig(zkUtils, topic, properties)
+
+                val configEntriesMap = config.entries.asScala.map(entry => (entry.name, entry.value)).toMap
+                policy.validate(new AlterConfigPolicy.RequestMetadata(
+                  new ConfigResource(ConfigResource.Type.TOPIC, resource.name), configEntriesMap.asJava))
+
+                if (!validateOnly)
+                  AdminUtils.changeTopicConfig(zkUtils, topic, properties)
```

Comment:
I agree it seems less than ideal, but restructuring to avoid it seems a bit annoying too. Since this should not be a heavily used API, I think it's probably OK as is.